File: CS_Algorithms/ADA_Algorithms/Greedy_Algorithms/C++/dijkstra.cpp
#include<iostream>
#include<limits>
#include<vector>
#include<algorithm>
#include<queue>
using namespace std; 
using pii = pair<int,int>;

class Graph{
    int V; vector<vector<pii>> adjList; 
public : 
    explicit Graph(int V): V(V){adjList.resize(V);}

    void addEdge(int src, int dest, int wt){
        adjList[src].push_back({dest, wt});
        adjList[dest].push_back({src, wt});
    }

    int getV() const { return V; }

    const vector<pii>& getNeighbors(int v) const{
        return adjList[v];
    }
};

class Dijkstra{
    const Graph& graph;
public:
    explicit Dijkstra(const Graph& g): graph(g){}
    pair<vector<int>,vector<int>>shortestPath(int src){
        const int inf = numeric_limits<int>::max();
        vector<int>dist(graph.getV(), inf);
        vector<int>parent(graph.getV(), -1);
        priority_queue<pii, vector<pii>, greater<pii>> pq;
        dist[src] = 0; pq.push({0, src});
        while(!pq.empty()){
            auto currentNode = pq.top(); pq.pop();
            int currentDist = currentNode.first;
            int currentV = currentNode.second;
            if(currentDist > dist[currentV]) continue;
            for(const auto& neighbor : graph.getNeighbors(currentV)){
                int nextV = neighbor.first;
                int weight = neighbor.second;
                int newDist = currentDist + weight;
                if(newDist < dist[nextV]){
                    dist[nextV] = newDist;
                    parent[nextV] = currentV;
                    pq.push({newDist, nextV});
                }
            }
        }
        return {dist, parent};
    }

    vector<int> reconstructPath(vector<int>&parent,int dest){
        vector<int>path; 
        if(parent[dest] == -1) return path;
        while(dest != -1){
            path.push_back(dest);
            dest = parent[dest];
        }
        reverse(path.begin(), path.end());
        return path;
    }
};


int main(){
    Graph g(7);
    g.addEdge(0,1,3); g.addEdge(1,2,7); 
    g.addEdge(2,4,15); g.addEdge(4,5,6); 
    g.addEdge(5,6,9); g.addEdge(6,1,14); 
    g.addEdge(1,3,9); g.addEdge(3,4,11); 
    g.addEdge(2,3,10); g.addEdge(6,3,2); 
 
    Dijkstra dijkstra(g);
    int src = 0 ; auto path = dijkstra.shortestPath(src);
    auto dist = path.first; auto parent = path.second;

    for(int i=0; i<dist.size(); i++){
        cout<<i << " "<< dist[i] << endl;
    }
    // Reconstructing path from source (0) to destination (5)
    int dest = 5;
    auto reconstructedPath = dijkstra.reconstructPath(parent, dest);
    for(int v : reconstructedPath){ cout << v << " ";}
    cout << endl; return 0 ; 
}

File: CS_Algorithms/ADA_Algorithms/Greedy_Algorithms/Python/dijkstra.py
import heapq
from typing import List, Tuple

class Graph:
    def __init__(self,vertices:int):
        self.V = vertices
        self.adjList = [[] for _ in range(vertices)]
    def addEdge(self,src:int,dest:int,wt:int):
        self.adjList[src].append((dest,wt))
        self.adjList[dest].append((src,wt))

class Dijkstra:
    def __init__(self,graph:Graph):
        self.graph = graph

    def shortestPath(self,src:int):
        dist = [float('inf')]*self.graph.V
        parent = [-1]*self.graph.V
        dist[src] = 0
        pq = [(0,src)]
        while pq:
            currentDist, currentV = heapq.heappop(pq)
            if currentDist > dist[currentV]:
                continue
            for neighbor, weight in self.graph.adjList[currentV]:
                distance = currentDist + weight
                if distance < dist[neighbor]:
                    dist[neighbor] = distance
                    parent[neighbor] = currentV
                    heapq.heappush(pq, (distance, neighbor))

        return dist, parent

    def reconstuctPath(self,parent:int,dest:int):
        path = []
        while dest != -1:
            path.append(dest)
            dest = parent[dest]
        return path[::-1]

def main():
    g = Graph(7);
    g.addEdge(0,1,3); g.addEdge(1,2,7); 
    g.addEdge(2,4,15); g.addEdge(4,5,6); 
    g.addEdge(5,6,9); g.addEdge(6,1,14); 
    g.addEdge(1,3,9); g.addEdge(3,4,11); 
    g.addEdge(2,3,10); g.addEdge(6,3,2); 
    dijkstra = Dijkstra(g)
    dist,parent = dijkstra.shortestPath(0)
    for v in range(g.V):
        print(f"Vertex : {v} : {dist[v]}")

    print("Path from 0 to 5:", dijkstra.reconstuctPath(parent,5))

if __name__ == "__main__":
    main()

